test-loop-block: run in qemu

bluca opened the pull request:
test-loop-block needs to run in qemu, so we are currently not
testing it in the CI. Run it by itself in a separate job from
TEST-02-UNITTESTS to avoid slowing that suite down.

Fixes https://github.com/systemd/systemd/issues/19966

The scripts are a copy of TEST-02-UNITTESTS, changing just to run under nspawn and to run the individual test case. We can add more as needed.

mbiebl commented:
hah, I guess it's a good thing that the crash in test-loop-block is reproducible :-)

bluca commented:
Interestingly the test only fails in our custom-built images, exact same binary works in a mkosi image. Output:

```
bash-5.1# /usr/lib/systemd/tests/test-loop-block
Fri 2021-06-18 15:19:33 BST (512) Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Fri 2021-06-18 15:19:33 BST (512) Found container virtualization none.
Checking that no-one is using this disk right now ... OK

Disk /var/tmp/.#sfdisk6719f57c9d709f0f: 256 MiB, 268435456 bytes, 524288 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

>>> Script header accepted.
>>> Created a new GPT disklabel (GUID: 49DEB8FB-442D-F94A-A80C-7B0C9A347DEC).
/var/tmp/.#sfdisk6719f57c9d709f0f1: Created a new partition 1 of type 'EFI System' and of size 32 MiB.
/var/tmp/.#sfdisk6719f57c9d709f0f2: Created a new partition 2 of type 'Linux extended boot' and of size 32 MiB.
/var/tmp/.#sfdisk6719f57c9d709f0f3: Created a new partition 3 of type 'Linux swap' and of size 32 MiB.
/var/tmp/.#sfdisk6719f57c9d709f0f4: Created a new partition 4 of type 'Linux root (x86-64)' and of size 32 MiB.
/var/tmp/.#sfdisk6719f57c9d709f0f5: Created a new partition 5 of type 'Linux home' and of size 32 MiB.
/var/tmp/.#sfdisk6719f57c9d709f0f6: Done.

New situation:
Disklabel type: gpt
Disk identifier: 49DEB8FB-442D-F94A-A80C-7B0C9A347DEC

Device                              Start    End Sectors Size Type
/var/tmp/.#sfdisk6719f57c9d709f0f1   2048  67583   65536  32M EFI System
/var/tmp/.#sfdisk6719f57c9d709f0f2  67584 133119   65536  32M Linux extended boo
/var/tmp/.#sfdisk6719f57c9d709f0f3 133120 198655   65536  32M Linux swap
/var/tmp/.#sfdisk6719f57c9d709f0f4 198656 264191   65536  32M Linux root (x86-64
/var/tmp/.#sfdisk6719f57c9d709f0f5 264192 329727   65536  32M Linux home

The partition table has been altered.








Syncing disks.

Fri 2021-06-18 15:19:33 BST (512) loop0: Waiting for device 'loop0' to initialize for 45s.
Fri 2021-06-18 15:19:33 BST (512) loop0: Successfully waited for device 'loop0' to initialize for 346us.
Fri 2021-06-18 15:19:33 BST (512) Image name .#sfdisk6719f57c9d709f0f is not valid, ignoring
Fri 2021-06-18 15:19:33 BST (512) No type detected on partition /dev/loop0p4
Fri 2021-06-18 15:19:33 BST (512) No type detected on partition /dev/loop0p5
Fri 2021-06-18 15:19:33 BST (512) No type detected on partition /dev/loop0p2
Fri 2021-06-18 15:19:33 BST (512) Successfully forked off '(mkfs)' as PID 523.
mkfs.fat 4.2 (2021-01-31)
Cannot initialize conversion from codepage 850 to ANSI_X3.4-1968: Invalid argument
Cannot initialize conversion from ANSI_X3.4-1968 to codepage 850: Invalid argument
Using internal CP850 conversion table
WARNING: Number of clusters for 32 bit FAT is less then suggested minimum.
Fri 2021-06-18 15:19:33 BST (512) (mkfs) succeeded.
Fri 2021-06-18 15:19:33 BST (512) Successfully forked off '(mkfs)' as PID 524.
mkfs.fat 4.2 (2021-01-31)
Cannot initialize conversion from codepage 850 to ANSI_X3.4-1968: Invalid argument
Cannot initialize conversion from ANSI_X3.4-1968 to codepage 850: Invalid argument
Using internal CP850 conversion table
WARNING: Number of clusters for 32 bit FAT is less then suggested minimum.
Fri 2021-06-18 15:19:33 BST (512) (mkfs) succeeded.
Fri 2021-06-18 15:19:33 BST (512) Successfully forked off '(mkfs)' as PID 525.
mke2fs 1.46.2 (28-Feb-2021)
Creating filesystem with 32768 1k blocks and 8192 inodes
Filesystem UUID: 03869920-cf3d-4962-b300-0974dee18447
Superblock backups stored on blocks: 
	8193, 24577

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (4096 blocks): done
Writing superblocks and filesystem accounting information: done

Fri 2021-06-18 15:19:33 BST (512) (mkfs) succeeded.
Fri 2021-06-18 15:19:33 BST (512) Successfully forked off '(mkfs)' as PID 526.
mke2fs 1.46.2 (28-Feb-2021)
Creating filesystem with 32768 1k blocks and 8192 inodes
Filesystem UUID: c257ef6a-9afa-4119-8ca0-322cbda89c51
Superblock backups stored on blocks: 
	8193, 24577

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (4096 blocks): done
Writing superblocks and filesystem accounting information: done

Fri 2021-06-18 15:19:33 BST (512) (mkfs) succeeded.
Fri 2021-06-18 15:19:33 BST (512) loop0: Waiting for device 'loop0' to initialize for 45s.
Fri 2021-06-18 15:19:33 BST (512) loop0: Successfully waited for device 'loop0' to initialize for 126us.
Fri 2021-06-18 15:19:33 BST (512) Image name .#sfdisk6719f57c9d709f0f is not valid, ignoring
Fri 2021-06-18 15:19:33 BST (512) Mounting /dev/loop0p4 (ext4) on /tmp/Sv6uPU (4 "")...
Fri 2021-06-18 15:19:33 BST (512) Mounting /dev/loop0p5 (ext4) on /tmp/Sv6uPU/home (4 "")...
Fri 2021-06-18 15:19:33 BST (512) Mounting /dev/loop0p2 (vfat) on /tmp/Sv6uPU/boot (4 "")...
Fri 2021-06-18 15:19:33 BST (512) Failed to mount /dev/loop0p2 (type vfat) on /tmp/Sv6uPU/boot (4 ""): Invalid argument
Fri 2021-06-18 15:19:33 BST (512) Assertion 'dissected_image_mount(dissected, mounted, UID_INVALID, UID_INVALID, 0) >= 0' failed at src/test/test-loop-block.c:220, function main(). Aborting.
Aborted (core dumped)
```

bluca commented:
I think this is the error from the kernel when mounting the vfat /boot partition:

`kernel: FAT-fs (loop0p2): codepage cp437 not found`

bluca commented:
two modules required by vfat.ko were missing, it works locally for me now

bluca commented:
@mrc0mmand any idea why `test-loop-block` would be skipped while running in qemu on CentOS?

mrc0mmand commented:
> @mrc0mmand any idea why `test-loop-block` would be skipped while running in qemu on CentOS?

I suspect this is the reason:

https://github.com/systemd/systemd-centos-ci/blob/a6db9b543166752dfb52d455de1370eda297f1f0/vagrant/test_scripts/test-arch.sh#L40-L46

bluca commented:
Ah yes indeed - great, so more flakiness expected here... I'll see how it goes, and might disable it again

keszybz reviewed (commented):
Duplicating all of this makes me wary. We had that before in the testsuite, and after a few years we end up with dozens of slightly different implementations. 

At least for the .sh scripts, it should be OK to either symlink to their `testsuite-02.*` counterparts, or maybe do something like
```sh
#!/bin/bin/env bash

. $(dirname $0)/testsuite-02.sh

...do some extra steps here if necessary...
```

The .service file should not be symlinked (because that creates issues with systemd thinking its an alias), and its short anyway.

bluca commented:
> Ah yes indeed - great, so more flakiness expected here... I'll see how it goes, and might disable it again

There we go:

```
Jun 21 16:56:46 H unknown[318]: Mon 2021-06-21 18:56:26 EEST (297) Failed dissect loopback device /dev/loop4: Connection timed out
```

Add it but keep it blocklisted? And hope we get the kernel fix before we retire?

bluca commented:
> Duplicating all of this makes me wary. We had that before in the testsuite, and after a few years we end up with dozens of slightly different implementations.
> 
> At least for the .sh scripts, it should be OK to either symlink to their `testsuite-02.*` counterparts, or maybe do something like
> 
> ```shell
> #!/bin/bin/env bash
> 
> . $(dirname $0)/testsuite-02.sh
> 
> ...do some extra steps here if necessary...
> ```
> 
> The .service file should not be symlinked (because that creates issues with systemd thinking its an alias), and its short anyway.

Ok, refactored a bit, and added to deny list.

mrc0mmand commented:
Hm, if I understand this correctly, `TEST-61-UNITTESTS-QEMU` is a subset of `TEST-02-UNITTEST`? Since `TEST-02` can run under both QEMU and nspawn (and at least in CentOS CI it does) and both tests use the same test set.

bluca commented:
> Hm, if I understand this correctly, `TEST-61-UNITTESTS-QEMU` is a subset of `TEST-02-UNITTEST`? Since `TEST-02` can run under both QEMU and nspawn (and at least in CentOS CI it does) and both tests use the same test set.

yeah it runs in nspawn in bionic, so test-loop-block is skipped

mrc0mmand reviewed (approved):
bionic-ppc64le failed in `test localed-locale` which is currently failing in other PRs as well, unrelated.

```
autopkgtest [20:46:01]: test localed-locale: [-----------------------
Generating locales (this might take a while)...
  en_US.UTF-8... done
Generation complete.
locale 'de_DE.UTF-8' not available
de_DE.UTF-8 UTF-8
locale 'de_DE.UTF-8' not available
de_DE.UTF-8 UTF-8
FAIL: expected: '0' actual: '1'
```
